File: egghead.io/data-structures-and-algorithms-in-js/lesson-09-tree-data-structure-in-js/tree.py
from node import Node


class Tree:
  def __init__(self, key):
    self.root = Node(key)

  def print(self):
    def traverse(node, func, depth):
      func(node, depth)

      if len(node.children):
        for child in node.children:
          traverse(child, func, depth + 1)
    
    def indent(node, depth):
      print(node.key 
        if depth == 0 
        else '{}{}'.format(' ' * depth, node.key)
      )
    
    traverse(self.root, indent, 0)

File: egghead.io/data-structures-and-algorithms-in-js/lesson-09-tree-data-structure-in-js/tree.js
const Node = require('./node');


class Tree {
  constructor(key) {
    this.root = new Node(key);
  }

  print() {
    function traverse(node, func, depth) {
      func(node, depth);

      if (node.children.length) {
        node.children.map(child => traverse(child, func, depth + 1));
      }
    }

    function indent(node, depth) {
      console.log(depth === 0
        ? node.key
        : `${' '.repeat(depth * 2)}${node.key}`
      );
    } 

    traverse(this.root, indent, 0);
  }
}

module.exports = Tree;
